File: dotnet/Zirve.Sdk/Cluster/ClusterManager.cs
using System;
using System.Net.Http;
using System.Text.Json;
using System.Threading.Tasks;
using Zirve.Sdk.Config;

namespace Zirve.Sdk.Cluster;

public class ClusterManager
{
    private readonly HttpClient _httpClient;
    private readonly string _url;
    private readonly string _token;

    public ClusterManager(HttpClient httpClient, ConfigManager configManager)
    {
        _httpClient = httpClient ?? throw new ArgumentNullException(nameof(httpClient));
        
        var cfg = configManager.Module("cluster");
        _url = (cfg.GetValueOrDefault("url", "http://rancher.cattle-system.svc.cluster.local") ?? "").TrimEnd('/') + "/v3";
        _token = cfg.GetValueOrDefault("token", "") ?? "";
    }

    private async Task<JsonElement> RequestAsync(HttpMethod method, string path, object? body = null)
    {
        var request = new HttpRequestMessage(method, $"{_url}/{path.TrimStart('/')}");
        request.Headers.Authorization = new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", _token);
        
        if (body != null)
        {
            var json = body is string s ? s : JsonSerializer.Serialize(body);
            request.Content = new StringContent(json, System.Text.Encoding.UTF8, "application/json");
        }

        var response = await _httpClient.SendAsync(request);
        
        if (response.StatusCode == System.Net.HttpStatusCode.NotFound)
            return default;

        response.EnsureSuccessStatusCode();

        if (response.StatusCode == System.Net.HttpStatusCode.NoContent || response.Content.Headers.ContentLength == 0)
            return default;

        using var doc = await JsonDocument.ParseAsync(await response.Content.ReadAsStreamAsync());
        return doc.RootElement.Clone();
    }

    public async Task<JsonElement> ListClustersAsync()
    {
        var res = await RequestAsync(HttpMethod.Get, "clusters");
        if (res.ValueKind == JsonValueKind.Object && res.TryGetProperty("data", out var data))
        {
            return data;
        }
        return res;
    }

    public async Task<JsonElement> GetClusterAsync(string clusterId)
    {
        return await RequestAsync(HttpMethod.Get, $"clusters/{Uri.EscapeDataString(clusterId)}");
    }

    public async Task<bool> IsHealthyAsync(string clusterId)
    {
        var cluster = await GetClusterAsync(clusterId);
        if (cluster.ValueKind == JsonValueKind.Object && cluster.TryGetProperty("state", out var state))
        {
            return state.GetString() == "active";
        }
        return false;
    }

    public async Task<bool> HealthAsync()
    {
        try
        {
            var response = await _httpClient.GetAsync(_url);
            return response.IsSuccessStatusCode;
        }
        catch
        {
            return false;
        }
    }
}

File: typescript/src/Cluster/ClusterManager.ts
/**
 * Zirve Cluster Manager — Rancher
 *
 * Interacts with Rancher API to list and query cluster statuses.
 */
export class ClusterManager {
  private config: Record<string, any>;
  private url: string;
  private token: string;

  constructor(config: Record<string, any>) {
    this.config = config;
    this.url = (config.url || 'http://rancher.cattle-system.svc.cluster.local').replace(/\/$/, '') + '/v3';
    this.token = config.token || '';
  }

  private async request(method: string, path: string, body?: any): Promise<any> {
    const options: RequestInit = {
      method,
      headers: {
        'Authorization': `Bearer ${this.token}`,
        'Content-Type': 'application/json'
      },
      body: body ? JSON.stringify(body) : undefined
    };

    const response = await fetch(`${this.url}/${path.replace(/^\/+/, '')}`, options);
    
    if (!response.ok) {
      if (response.status === 404) return null;
      throw new Error(`Rancher API Error [${response.status}]: ${response.statusText}`);
    }

    if (response.status === 204 || response.headers.get('content-length') === '0') return true;
    return await response.json();
  }

  public async listClusters(): Promise<any[]> {
    const result = await this.request('GET', 'clusters');
    return result?.data || [];
  }

  public async getCluster(clusterId: string): Promise<any | null> {
    return await this.request('GET', `clusters/${clusterId}`);
  }

  public async isHealthy(clusterId: string): Promise<boolean> {
    const cluster = await this.getCluster(clusterId);
    return cluster && cluster.state === 'active';
  }

  public async health(): Promise<boolean> {
    // Rancher v3 API root returns configuration state
    try {
      const response = await fetch(`${this.url}`);
      return response.ok;
    } catch {
      return false;
    }
  }
}
